ssl/record/rec_layer_s3.c

```diff
@@ -1231,7 +1231,7 @@ int ssl3_write_pending(SSL *s, int type, const unsigned char *buf, size_t len,
                      SSL_R_BIO_NOT_SET);
             i = -1;
         }
-        if (i > 0 && tmpwrit == SSL3_BUFFER_get_left(&wb[currbuf])) {
+        if (i >= 0 && tmpwrit == SSL3_BUFFER_get_left(&wb[currbuf])) {
```

Comment:
It looks like we can only get here with `i==0` if `BIO_write()` returned 0, but that state is ambiguous as to "successful write of 0 bytes" vs "unsuccessful write", and it's not actually clear that switching to `BIO_write_ex()` fixes that (since the underlying `bwrite` methods basically have the same "number of bytes written" semantics for return values).
That said, the `tmpwrit` check should be enforcing that we actually attempted a 0-byte write, so I'm inclined to say that this is the right thing to do...but it may need some additional pondering.

Reply:
I don't understand what you are trying to achieve from this change? Under what circumstances should `i==0` be counted as success here? We shouldn't even get here if we have 0 bytes to write?

Reply:
To be clear, the scenario in question is supporting empty fragments for TLS 1.0.  With KTLS, the kernel is responsible for adding the TLS header and trailer, so an empty fragment is a literal write() or sendmsg() of 0 bytes.  In the current code, when OpenSSL attempts to send the empty fragment via the recursive call to `do_ssl3_write`, the payload length (`tmppipelen`) is 0.  The KTLS path bypasses the code to add the TLS header and footer (the `goto wpacket_init_complete` and various `BIO_get_ktls_send` checks in subsequent code paths), so the resulting `totlen` passed to `ssl3_write_pending` is 0.

The first patch in this series is the one-line change above which treats the 0 returned by write() inside of bss_sock.c's write method as success in the case that the write size was 0.  However, I wasn't sure if that would break any other use cases, so the second patch tries a more targeted approach specifically for empty fragments in the KTLS case of avoiding the recursion into `do_ssl3_write` and instead using `BIO_set_ctrl_ktls_msg` and `BIO_write` explicitly to trigger a sendmsg() of a app-data TLS record with a 0 byte payload.

I've tested both patches with TLS 1.0 and KTLS in FreeBSD, so I am fine with whichever approach is preferred.

Reply:
(Just to state it again for Matt, my preference was to take this one and not bother with the KTLS-specific stuff.)

Reply:
Aha! Right now I understand it. Thanks for the explanation. I agree with @kaduk, this simpler solution seems to be the way to go. I wonder if it is worthy of a comment somewhere around here explaining that in KTLS CBC records we can legitimately attempt to write 0 bytes.

Another solution (but more work) would be to convert `BIO_s_connect`/`BIO_s_socket` to use the "new style" write functions such that we could use BIO_write_ex here and get reliable results.

Reply:
Ok, I've dropped the second change and added a comment to this change.  I've also adjusted the commit log per @kaduk 's suggestion.

I'm not opposed to the larger solution and might try to tackle it as a separate patch, but I believe it would need interleaving BIO's to also support the *_ex methods, e.g. the buffer filter (bf_buff.c I believe?) would also have to be updated as when I've debugged KTLS before I found that the socket BIO's were nested underneath a buffer BIO.  It seems like it might be ideal to switch as many BIO backends over to the new method as possible due to this, though perhaps getting the "filter" BIOs such as buffer cut over first would be most important?  I'm not really sure and would prefer to do that separately as a result.